File: interviewPython/interviewpython/rotate_array.py
class roateArray_solution(object):
    def rotateArray(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        """
        n = len(nums)
        if (n == 0 or k == 0):
            return
        self.reverse(nums, 0, n-k-1)
        self.reverse(nums, n-k, n-1)
        self.reverse(nums, 0, n-1)
        
    def reverse(self, nums, l, r):
        while(l < r):
            tmp     = nums[l]
            nums[l] = nums[r]
            nums[r] = tmp
            l = l + 1
            r = r - 1


if __name__ == '__main__':
    print("PY:rotateArray")
    solu = roateArray_solution()
    s = [1,2,3,4,5,6,7]
    solu.rotateArray(s, 3)
    print(s)
            

File: interviewCode/src/others/rotatearray/rotatearray.cpp
/*
 * rotatearray.cpp
 *
 *  Created on: Nov 27, 2016
 *      Author: songjiguo
 */

/*
Rotate an array of n elements to the right by k steps.
For example, with n = 7 and k = 3, the array [1,2,3,4,5,6,7] is rotated to [5,6,7,1,2,3,4].
*/

#include "rotatearray.h"

static void revserse(vector<int> &nums, int left, int right)
{
	while(left < right) {
		int tmp = nums[left];
		nums[left] = nums[right];
		nums[right] = tmp;
		++left;
		--right;
	}
}

void rotatearray::rotate(vector<int> &nums, int k)
{
	if(k <= 0 || nums.empty()) return;
	k = k%nums.size();

	revserse(nums, 0, nums.size() - k - 1);
	revserse(nums, nums.size()-k, nums.size() - 1);
	revserse(nums, 0, nums.size() - 1);
}
